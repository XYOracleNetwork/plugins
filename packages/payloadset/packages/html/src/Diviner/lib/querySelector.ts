import { exists } from '@xylabs/exists'
import { load } from 'cheerio'

const opts = {}
// const opts = { decodeEntities: false }

/**
 * Takes a string data field containing html and converts it to an XML object
 * @param html The input object which contains the stringified html data
 * @returns A promise that resolves to an XML payload
 */
export const querySelector = (html: string, selector: string): string | null => {
  // Check if the input string is empty or null
  if (!html.trim()) return null
  // Parse the HTML
  const dom = load(html, opts)
  const element = dom(selector)
  // TODO: .toString() instead of .html()?
  return element.html()
}

export const querySelectorAll = (html: string, selectors: string): string[] => {
  // Check if the input string is empty or null
  if (!html.trim()) return []
  // Parse the HTML
  const dom = load(html, opts)
  const elements = dom(selectors) as unknown as ReturnType<typeof dom>[]
  return elements.map(element => element.html()).filter(exists)
}
