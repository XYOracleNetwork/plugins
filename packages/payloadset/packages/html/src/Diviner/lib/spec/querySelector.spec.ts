import { describe, expect, it } from 'vitest'

import { querySelector } from '../querySelector.ts'

describe('querySelector', () => {
  describe('returns the first matching element from the supplied HTML', () => {
    const html = `
    <div id="main" class="container">
      <h1 class="title">Hello, World!</h1>
      <p class="description">This is a test paragraph.</p>
      <a href="/link" class="link">Click Here</a>
      <ul>
        <li class="item">Item 1</li>
        <li class="item">Item 2</li>
        <li class="item">Item 3</li>
      </ul>
      <input type="text" name="username" value="testUser" />
      <img src="image.png" alt="Test Image" />
    </div>
  `
    const cases = [
      ['should find element by tag name', 'h1'],
      ['should find element by class name', '.description'],
      ['should find element by id', '#main'],
      ['should find multiple elements by class name', '.item'],
      ['should find element by attribute', 'input[name="username"]'],
      ['should find element by attribute with specific value', 'img[src="image.png"]'],
      ['should find element by attribute starts with', 'a[href^="/"]'],
      ['should find element by class and tag name combination', 'h1.title'],
      ['should find element by tag name inside another element', '#main h1'],
    ]
    it.each(cases)('%s', (_, selector) => {
      const result = querySelector(html, selector)
      expect(result).not.toBeEmpty()
      expect(result).toMatchSnapshot()
    })
  })
  describe('returns null', () => {
    it('with empty selector', () => {
      const result = querySelector('<div><p>Test</p></div>', '')
      expect(result).toBeNull()
    })
    it('with empty html', () => {
      const result = querySelector('', '*')
      expect(result).toBeNull()
    })
    it('with empty selector and empty html', () => {
      const result = querySelector('', '')
      expect(result).toBeNull()
    })
    it('with no matching data', () => {
      const result = querySelector('<div><p>Test</p></div>', 'h1')
      expect(result).toBeNull()
    })
  })
})
