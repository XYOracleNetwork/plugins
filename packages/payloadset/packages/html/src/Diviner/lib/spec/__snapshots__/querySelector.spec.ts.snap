// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`querySelector > returns the first matching element from the supplied HTML > should find element by attribute 1`] = `""`;

exports[`querySelector > returns the first matching element from the supplied HTML > should find element by attribute starts with 1`] = `"Click Here"`;

exports[`querySelector > returns the first matching element from the supplied HTML > should find element by attribute with specific value 1`] = `""`;

exports[`querySelector > returns the first matching element from the supplied HTML > should find element by class and tag name combination 1`] = `"Hello, World!"`;

exports[`querySelector > returns the first matching element from the supplied HTML > should find element by class name 1`] = `"This is a test paragraph."`;

exports[`querySelector > returns the first matching element from the supplied HTML > should find element by id 1`] = `
"
      <h1 class="title">Hello, World!</h1>
      <p class="description">This is a test paragraph.</p>
      <a href="/link" class="link">Click Here</a>
      <ul>
        <li class="item">Item 1</li>
        <li class="item">Item 2</li>
        <li class="item">Item 3</li>
      </ul>
      <input type="text" name="username" value="testUser">
      <img src="image.png" alt="Test Image">
    "
`;

exports[`querySelector > returns the first matching element from the supplied HTML > should find element by tag name 1`] = `"Hello, World!"`;

exports[`querySelector > returns the first matching element from the supplied HTML > should find element by tag name inside another element 1`] = `"Hello, World!"`;

exports[`querySelector > returns the first matching element from the supplied HTML > should find multiple elements by class name 1`] = `"Item 1"`;

exports[`querySelector > returns the first matching element from the supplied HTML > with valid selector and matching data 1`] = `null`;

exports[`querySelector > returns the first matching element from the supplied HTML > with valid selector and matching data 2`] = `null`;

exports[`querySelector > returns the first matching element from the supplied HTML > with valid selector and matching data 3`] = `null`;
